Add Waf plugin (LP: #1611335)

cpaelzer opened the pull request:
Hi,
I was snapping a project using the waf build sytem (https://github.com/waf-project/waf) and realized snapcraft has no waf plugin yet. This is added by this PR together with a unittest for it.

snappy-m-o commented:
Can one of the admins verify this patch?

snappy-m-o commented:
Can one of the admins verify this patch?

cpaelzer commented:
Hmm, all local tests ran fine - is there something I can/should do to please travis?

sergiusens commented:
I retriggered the unit task as it seems to have failed on a network error when `pip install`ing the deps.

sergiusens commented:
and there are static issues:
```
snapcraft/tests/test_plugin_waf.py:18:1: F401 'stat' imported but unused
snapcraft/tests/test_plugin_waf.py:94:44: E127 continuation line over-indented for visual indent
snapcraft/tests/test_plugin_waf.py:18:1: F401 'stat' imported but unused
snapcraft/tests/test_plugin_waf.py:94:44: E127 continuation line over-indented for visual indent
```

fwiw we are still using flake8 from the xenial archives and not from pypi (there's a task to move that).

cpaelzer commented:
Thanks sergiusens for taking a look.
A local runtest didn't complain before I submitted - but that might as well be caused by the "wrong" flake8 used.

Anyway I switched on all checks in my syntastic plugin and addressed all you mentioned and a few more. 
There are only a few pylint warnings left, but that is known to be too-noisy sometimes (and for these warnings almost all of snapcraft would have to be rewritten like super-on-old-class).

Updating the PR now.

sergiusens commented:
Ok looks good. Do you have a snapcraft.yaml ready that would build with this?
If it is not too much to ask, do yo have something small enough (hello world type) that could be added to `integrations_tests/snaps` and hook it up in `integration_tests/test_waf_plugin.py` (new file)

cpaelzer commented:
I have an example snapcraft.yaml here (I still provide it as a custom plugin there for now):
https://gitlab.com/paelzer/ntpsec/blob/snapify-ntpsec/snapcraft.yaml

I have nothing, but I'll give it a try to snap something minimal for the integration tests.
I'll notify you on this PR then, if it is done or if there were and blockers stopping me.

cpaelzer commented:
The integration test I added is based on the demos waf is providing by themselves.
That way it is already having a few special cases that will be covered by the integration tests.

There is no packaged waf yet, and actually projects using it carry it in their git as far as I could see, think of a configure script. So does the integration test for now based on latest waf 1.9.2

The only things that could bite us with that are:
- waf has a BSD license, so has the integrated .waf script
- the integrated script is generated and gets a lot of py checkers upset, as far as I could see in runtests.sh we are not checking the integration snap examples thou so that should be good.
- the test program has no a lot of output to check, I could only check RC but that is currently not supported in the tests (see comments at get_output_ignoring_non_zero_exit). So I chose (like maven) just to run the build itself.

The test runs locally fine e.g. isolated from the rest for fast eval via:
`python3 -m unittest discover -b -v -s integration_tests -p 'test_waf_plugin.py'`
Please let me know if it does so for you as well inside the "real" CI chain.

sergiusens commented:
Hi! Mind fixing the unit tests?

cpaelzer commented:
Hi,
No I don't mind :-)
Just was on PTO - taking a look now why it failed.

cpaelzer commented:
I found it, the test was broken due to another upstream commit. I merged latest upstream to find it and added a fix. Pushed it again and hope that everything works now and is ready to be merged.

Locally tests were good again now after the fix, but we will have to wait if travis agrees :-)

sergiusens commented:
@cpaelzer hey, it looks like something went wrong with your merge/rebase. Mind fixing it?

cpaelzer commented:
@sergiusens hi, to me it looks all good. All tests passed and it is reported to be able to be automatically mergeable.

I can only think of - instead of having a merge commit in my branch that I think github should handle well - instead rebasing it with git rebase, but that would mean a new branch or a force push. Before I do either of that I'd ask you what actually went wrong as I can't identify the issue on my end.

cpaelzer commented:
I realized we are off a few timezones and didn't want to loose time.
So to "clean" it up in general I fetched from upstresm and rebased my master which as expected was a clean fast forward. To then rebase onto that new master.
Also as expected that needed a force push to get the old merge commit out of the way but it looks cleaner now.

cpaelzer commented:
After waiting a bit I see the tests are good again as well.
Please if still something is broken from your POV go into some detail what I should fix :-)

elopio commented on integration_tests/snaps/simple-waf/program/wscript_build:
```diff
@@ -0,0 +1,39 @@
+
+# ----------------------------------------
+
+from waflib import TaskGen
```
a nit, can this be on the top of the file?

cpaelzer replied:
This as well as the other is not written from scratch - as I wrote in the commit the whole simple-waf program is from upstream waf demo section as-is.
I surely can move it up, but I'd rather prefer it pristine as provided by the waf project. That lessens the likelihood of somebody wondering why it is different than upstream and will easen if we ever have to update it as any upstream diff will apply

elopio commented on integration_tests/snaps/simple-waf/program/wscript_build:
```diff
@@ -0,0 +1,39 @@
+	tsk = self.create_task('foo')
+	tsk.outputs = [self.bld.path.find_or_declare('abc.h')]
+
+import waflib.Task
```
on the top preferably, too.

cpaelzer replied:
As above, it could be done - but without a "hard order" I'd prefer not to for the reasons given above.

elopio commented on integration_tests/snaps/simple-waf/waf:
```diff
@@ -0,0 +1,169 @@
+STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
+IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+"""
```
this needs a notice in debian/copyright, right @sergiusens ?

elopio commented on integration_tests/snaps/simple-waf/waf:
```diff
@@ -0,0 +1,169 @@
+		return x.encode()
+
+def err(m):
+	print(('\033[91mError: %s\033[0m' % m))
```
use .format instead of %, please.

elopio replied:
oh, this is a file copied from upstream. Nevermind, we don't want to diverge.
However this uses tabs instead of spaces, ugh :)

elopio commented:
Hello @cpaelzer! Thanks a lot for your contribution.
What do you think about making a waf part in the wiki instead of putting it in our integration test?

cpaelzer commented:
Since I won't have time now a little inline doc here about what to do:

[17:04] <elopio> cpaelzer: hello. Have you seen https://wiki.ubuntu.com/snapcraft/parts ? You can put in there a part to be reused. I thought waf would be a good case for that.
[17:04] <elopio> instead of copying the waf source in your snap, you would use: after: [waf], and snapcraft will bring it from the wiki.
[17:07] <cpaelzer> elopio: I've seen that in the past, but it appeared to me a bit too "unofficial" and since I want to get an external project to use snaps I'd prefer to have waf "properly" in snapcraft
[17:08] <cpaelzer> elopio: I might not see the whole of it - could I use that mechanism to "just keep the integration test" in such a part, but let the rest be in "official" snapcraft?
[17:15] <elopio> cpaelzer: the wiki is pretty official. You would keep the waf plugin in snapcraft, and a waf part in the wiki. The part will just contain the waf python files you are putting in the integration test.
[17:18] <cpaelzer> elopio: uh I see so it would only hold the extra stuff I've moved from the waf upstream into the integration test and that way "stay out of tree"
[17:18] <cpaelzer> elopio: I guess the integration test then is the one using the after [waf] - yeah if that works I'm good
[17:18] <cpaelzer> elopio: I'm giving it a try til the next meeting starts
[17:18] <elopio> cpaelzer: yes, thanks.

I'll take a look and give an update here.

cpaelzer commented:
Moved the waf demo bits to remote part as discussed now - thanks to @josepht for clarifying my remote part issues.
Rebased to latest master again and ran unit/integration tests again (some usually fail locally, but none related to the PR).

I hope the tests pass on the CI as well and this PR is then ready for merge consideration.
Kind Regards,
Christian

elopio commented:
Hey @cpaelzer, I like this very much! Sadly, during today's standup we discussed about it and found that we are lacking some important features to use this style of plugins with dependencies as parts. For example, the plugin will depend on the part and that would have to be manual as there's no way in the base plugin class to specify it, so it's likely to be forgotten. And the plugin will depend on your git branch, so we need a way to pin or vendor the dependencies. It was a great experiment, I'm just sorry to have requested it in your branch and made you do all the work instead of doing a prototype myself first.
We decided that for now the plugin must install all the dependencies, and that we will work on the missing bits to use parts in a future version. Can I ask you to revert to the previous state and bundle waf in snapcraft as part of the plugin? :worried: I suck I know, I owe you one. :worried:
But on the biright side, we now have a good idea of what we want. What are your thoughts? Did you like using the remote parts?

cpaelzer commented:
Hi @elopio - of course I can revert the change to make it a remote part. Don't be afraid I hate you only just a little now :-P

No honestly it is fine. It was an effort worth taking for my personal education to see the remote parts. I think that on this end of snap* there is a lack of documentation of workflow and examples. Other than this initial hurdle it was ok after I kind of "got the point" - in fact I like the ability to have remote parts in general.

Eventually it was straight forward, but almost everything looks like that when looking back. I'd expect them to have more target specifiers than the wiki like git/tar/file/... as e.g. a source attribute, but maybe they already have and I just didn't check that as an option yet.

sergiusens commented:
ok to test

sergiusens commented:
I don't want to be that guy, but maybe consider downloading waf as part of the *pull* step, I see `waf` offered here https://waf.io/ and you can implement similar logic to what we have in the nodejs plugin to retrieve nodejs/npm itself.

cpaelzer commented:
I'm currently and likely til Yakkety is out super-busy with plenty of other
things.
If it would be possible to accept as-is, but open and assign a bug to me to
later improve in that regard that would be highly welcome by me.

Please discuss in your Team and let me know if that would ok or if you need
it right away for the PR to land.

On Wed, Aug 31, 2016 at 7:51 PM, Sergio Schvezov <notifications@github.com>
wrote:

> I don't want to be that guy, but maybe consider downloading waf as part of
> the *pull* step, I see waf offered here https://waf.io/ and you can
> implement similar logic to what we have in the nodejs plugin to retrieve
> nodejs/npm itself.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/snapcore/snapcraft/pull/716#issuecomment-243845243>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AOTJnm-Bp2p0jbIfpLJodUshItxrYOVmks5qlb8lgaJpZM4JgFLo>
> .
>



-- 
Christian Ehrhardt
Software Engineer, Ubuntu Server
Canonical Ltd

cpaelzer commented:
Hi - finally found a few minutes.

The nodejs/npm solution is for something that has defined packages. waf is usually bundled with the project that is compiled and becoming the snap.

The discussion here is circling around the waf test which so far either pulled from wiki as remote part or had a local testcase (taken from waf.io) with plenty of python code - both unwanted options.

But the testcase should resemble a real project that uses waf and that has waf integrated just as old projects have a configure script. There is no point in adding a "fetch waf" feature to the waf plugin. Even less a feature that fetches the demo we need only for this test.

I know that the book https://waf.io/book/#_waf_commands_and_usage defines only wscript as "project-specific details are best kept and versioned in files residing along with the project source code. These files are modules written in the Python programming language and are named wscript.".
But the project using was I found so far was bundling waf itself as well to avoid breakage on unexpected version changes. So pulling the latest and greatest waf by-default might break all of them.

The only thing left that could be done might be to to pull from waf.io git and execute the test on that. But that pulls way more than needed and it would make you dependent on another project. If they break themselve the snapcraft integration tests would break. I don't think that is what you want - right?

Maybe we could pull in latest waf from waf.io IF there is no local bundled waf. Give the plugin a parm that can force this to always pull it in, but be default off. That way the demo could drop the waf script but would have to keep everything else.
But this gets more and more complex and I'm not convinced to do so just on an assumption.

Since I waited for your decision before I just rebased for now to get rit of the merge conflicts.
Pushing the rebase now...

I'd really like to get it accepted as-is as working V1 of the waf plugin. I don't mind if I or anybody else adds the "pull in waf if not bundled" feature later, but I don't see it as required part of the initial waf plugin support.
If inacceptable please catch me on (irc cpaelzer) for a more in depth discussion on the alternative that you'd like to see implemented here - I don't want to head in the wrong direction based on too many assumptions.

sergiusens commented:
@cpaelzer given everything you've mentioned, we should keep the PR with how you originally wrote the integration test.
I believe this is the case as the PR is now, I will give it a light review and if all is good, I will merge it and fix `debian/copyright` (if you do not want to yourself) to take into account the licensing in the waf specific files.

cpaelzer commented:
The upstream project that I was trying to snap are short of releasing their 1.0
Once they do so they will have created usual deb/rpm packaging and such.
I initially proposed snaps for this and they were kind of open, but I feel we are already starting to slip the chance to get making it a snap.

I'll still miss the interfaces I need as I only had time to open bugs for them but not provide PRs on my own (1616833/1616834). But at least having the build system plugin not being part of the PR for the upstream project would make it appear more mature.

That said - any update on integrating this PR since last week?
